### PR TITLE
Be a little more careful when removing mount points.

### PR DESCRIPTION
**Pull request by brendandburns**

### CODE REVIEW

**Comment by brendandburns** [comment]
@thockin

File: pkg/kubelet/volume/gce_pd/gce_pd.go [hunk 1]
```diff
@@ -194,9 +194,20 @@ func (pd *gcePersistentDisk) SetUp() error {
...
-		os.RemoveAll(pd.GetPath())
-		// TODO: we should really eject the attach/detach out into its own control loop.
-		detachDiskLogError(pd)
+		mountpoint, mntErr := isMountPoint(pd.GetPath())
+		if mntErr != nil {
+			glog.Errorf("Failed to mountpoint check: %v", mntErr)
+			return err
+		}
+		if !mountpoint {
+			os.RemoveAll(pd.GetPath())
```
Comment (thockin) [inline pkg/kubelet/volume/gce_pd/gce_pd.go]: Worth a sanity check at the second RemoveAll in TearDown?

File: pkg/kubelet/volume/gce_pd/gce_pd.go [hunk 2]
```diff
@@ -194,9 +194,20 @@ func (pd *gcePersistentDisk) SetUp() error {
 	globalPDPath := makeGlobalPDName(pd.plugin.host, pd.pdName, pd.readOnly)
 	err = pd.mounter.Mount(globalPDPath, pd.GetPath(), "", mount.FlagBind|flags, "")
 	if err != nil {
-		os.RemoveAll(pd.GetPath())
-		// TODO: we should really eject the attach/detach out into its own control loop.
-		detachDiskLogError(pd)
+		mountpoint, mntErr := isMountPoint(pd.GetPath())
+		if mntErr != nil {
+			glog.Errorf("Failed to mountpoint check: %v", mntErr)
```
Comment (thockin) [inline pkg/kubelet/volume/gce_pd/gce_pd.go]: Weird error message

**Comment by thockin** [comment]
LGTM mod nits

**Comment by satnam6502** [comment]
@brendandburns : I think we just need the nits address before merging? Thanks.

File: pkg/kubelet/volume/gce_pd/gce_pd.go [hunk 3]
```diff
@@ -194,9 +194,20 @@ func (pd *gcePersistentDisk) SetUp() error {
...
 	if err != nil {
-		os.RemoveAll(pd.GetPath())
-		// TODO: we should really eject the attach/detach out into its own control loop.
-		detachDiskLogError(pd)
+		mountpoint, mntErr := isMountPoint(pd.GetPath())
+		if mntErr != nil {
+			glog.Errorf("isMountpoint check failed: %v", mntErr)
+			return err
+		}
+		if !mountpoint {
```
Comment (thockin) [inline pkg/kubelet/volume/gce_pd/gce_pd.go]: I'd rather see this structured a bit more obviously:

```
... = isMountPoint()
if mntErr {  log it & return }
if mountpoint {
    call unmount
    ... = isMountPoint() // again
    if mountpoint {
        // Comment saying this is very odd, and we'll try again next sync loop
        log(this is bad, we'll try again next sync loop)
        return
    }
    os.Remove()
    detach
```

Specifically the comment  that we will try again is important to comprehension.

**Comment by thockin** [comment]
Just 1 comment, the LGTM

File: pkg/kubelet/volume/gce_pd/gce_pd.go (hunk 3 above)
Comment (brendandburns) [inline pkg/kubelet/volume/gce_pd/gce_pd.go]: done.

**Comment by brendandburns** [comment]
Comment addressed.  This code is still really convoluted.  I'm going to take a stab at making it cleaner in a different PR.

File: pkg/kubelet/volume/gce_pd/gce_pd.go [hunk 4]
```diff
@@ -194,9 +194,31 @@ func (pd *gcePersistentDisk) SetUp() error {
...
+			if mntErr != nil {
+				glog.Errorf("isMountpoint check failed: %v", mntErr)
+				return err
+			}
+			if mountpoint {
+				// This is very odd, we don't expect it.  We'll try again next sync loop.
+				glog.Errorf("%s is still mounted, despite call to unmount().  Will try again next sync loop.", pd.GetPath())
+				return err
+			}
+		} else {
```
Comment (thockin) [inline pkg/kubelet/volume/gce_pd/gce_pd.go]: This should not be an else, just move to to parent scope.  The only way to arrive here is if isMountPoint() returned false.

**Comment by brendandburns** [comment]
Comment addressed.  ptal.

Thanks
--brendan

**Pull request merged**